lib/active_merchant/billing/gateways/redsys_rest.rb

```diff
@@ -338,14 +341,23 @@ def add_order(post, order_id)
         post[:DS_MERCHANT_ORDER] = clean_order_id(order_id)
       end
 
-      def add_payment(post, card)
-        name = [card.first_name, card.last_name].join(' ').slice(0, 60)
-        year = sprintf('%.4i', card.year)
-        month = sprintf('%.2i', card.month)
-        post['DS_MERCHANT_TITULAR'] = CGI.escape(name)
-        post['DS_MERCHANT_PAN'] = card.number
-        post['DS_MERCHANT_EXPIRYDATE'] = "#{year[2..3]}#{month}"
-        post['DS_MERCHANT_CVV2'] = card.verification_value if card.verification_value.present?
+      def add_payment(post, payment_method)
+        year = sprintf('%.4i', payment_method.year)
+        month = sprintf('%.2i', payment_method.month)
+
+        if payment_method.is_a?(NetworkTokenizationCreditCard)
+          post[:Ds_Merchant_TokenData] = {
+            tokenCryptogram: payment_method.payment_cryptogram,
+            expirationDate: "#{year[2..3]}#{month}",
+            token: payment_method.number
+          }
+        else
+          name = [payment_method.first_name, payment_method.last_name].join(' ').slice(0, 60)
+          post['DS_MERCHANT_TITULAR'] = CGI.escape(name)
```

Comment:
just out of curiosity do you know why is needed escape special characters on this value ?

Reply:
Not sure, It could be to treat the possible special characters, the tests seams to work, and I didn't implement that part and can't found a reason.